### PR TITLE
Migration

### PR DESCRIPTION
**Pull request by henryzbw**

### CODE REVIEW

File: src/db/migrations/migration2.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+          check => check(null, i < txs.length), // trigger iter
+          (next) => {
+            const tx = txs[i];
+            i++;
+            try {
+              const userAddress = tx.betterAddress || tx.centralizedOracleAddress;
+              const eventLeaderboardEntry = new EventLeaderboard({
+                eventAddress,
+                userAddress,
+                investments: tx.amount, // event leaderboard entry already has user's investments
```
Comment (BernardHan) [inline src/db/migrations/migration2.js]: remove the copied comment, it's misleading in this migration

File: src/db/migrations/migration2.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        txs.forEach((tx, index) => {
+          if(tx.resultIndex !== event.currentResultIndex) return;
+          if(tx.betterAddress) addresses.add(tx.betterAddress);
+          if(tx.centralizedOracleAddress) addresses.add(tx.centralizedOracleAddress);
+        })
+        addresses.forEach((address, index) => {
+          promises.push(limit(async () => {
+            try {
+              // calculate winning for this user in this event
+              let winnings = '0';
```
Comment (BernardHan) [inline src/db/migrations/migration2.js]: assign winnings after await with `const`, it never gets reassigned

**Comment by BernardHan** [comment]
@dwalintukan may want the `plimit` to wrap the very outer logic, but migration is one time execution, if this does not cause problem, I'd say it looks good to me.

File: src/db/migrations/migration2.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+      }
+      if (event.status === EVENT_STATUS.WITHDRAWING) {
+        const addresses = new Set();
+        txs.forEach((tx, index) => {
+          if(tx.resultIndex !== event.currentResultIndex) return;
+          if(tx.betterAddress) addresses.add(tx.betterAddress);
+          if(tx.centralizedOracleAddress) addresses.add(tx.centralizedOracleAddress);
+        })
+        addresses.forEach((address, index) => {
+          promises.push(limit(async () => {
```
Comment (dwalintukan) [inline src/db/migrations/migration2.js]: move this `limit` call to the very outer loop. so 1 promise will take care of 1 event.

this outer loop `events.forEach(async (event, key) => {` will keep running even though the `limit` here will pause operation so it "should" queue everything up anyways once it hits the 15 concurrency limit. prefer to apply the `limit` on the entire event parsing and have each `calculateWinnings` call sequentially. so each `limit` func will correspond to all the inserts + calcWinnings of one event, and calcWinnings will only be called once at a time per event. 

1. move `limit` to very outer loop
2. use `eachOfSeries` in async lib to handle calculateWinnings

File: src/db/migrations/migration2.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        eventAddress: event.address,
+        txStatus: TX_STATUS.SUCCESS,
+      });
+      const txs = resultSet.concat(bets);
+
+      try {
+        let i = 0;
+        await async.whilst(
+          check => check(null, i < txs.length), // trigger iter
+          (next) => {
```
[the remark this answers is not in the document]
Comment (henryzbw) [reply src/db/migrations/migration2.js]: tried declare as async there, will throw error as next is not a function

File: src/db/migrations/migration2.js [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+          eventAddress: event.address,
+          txStatus: TX_STATUS.SUCCESS,
+        });
+        const txs = resultSet.concat(bets);
+
+        try {
+          let i = 0;
+          await async.whilst(
+            check => check(null, i < txs.length), // trigger iter
+            (next) => {
```
Comment (dwalintukan) [inline src/db/migrations/migration2.js]: Rename next, this is shadowing the same var as the `async function migration2(next) {`.

File: src/db/migrations/migration2.js [hunk 6]
```diff
@@ -0,0 +1,120 @@
+const pLimit = require('p-limit');
+const async = require('async');
```
Comment (dwalintukan) [inline src/db/migrations/migration2.js]: nit: destructure only funcs that you need

File: src/db/migrations/migration2.js (hunk 5 above)
Comment (dwalintukan) [inline src/db/migrations/migration2.js]: I just tested, you can use `async () => {`. See here for how to implement: https://caolan.github.io/async/v3/global.html

```
await async.whilst(
            check => check(null, i < txs.length), // trigger iter
            async () => {
              const tx = txs[i];
              i++;
              const userAddress = tx.betterAddress || tx.centralizedOracleAddress;
              const eventLeaderboardEntry = new EventLeaderboard({
                eventAddress,
                userAddress,
                investments: tx.amount,
                winnings: '0',
              });
              await DBHelper.insertEventLeaderboard(eventLeaderboardEntry);
              if (event.status === EVENT_STATUS.WITHDRAWING) {
                const globalLeaderboardEntry = new GlobalLeaderboard({
                  userAddress,
                  investments: tx.amount,
                  winnings: '0',
                });
                await DBHelper.insertGlobalLeaderboard(globalLeaderboardEntry);
              }
            },
          );
```

**Pull request merged**